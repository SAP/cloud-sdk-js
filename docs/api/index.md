# API Documentation

- [Version 1.18.1](1.18.1),
- [Version 1.18.0](1.18.0)