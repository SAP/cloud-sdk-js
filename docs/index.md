# SAP Cloud SDK Documentation

## API Reference
The API reference is to be found [here](api).
