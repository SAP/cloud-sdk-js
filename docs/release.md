# Release Process
All SAP Cloud SDK modules will be published with the same version regardless whether there where changes within the particular packages or not.

## How to trigger a release
The release process can only be triggered by owners of the repository. It is triggered by bumping the version using `lerna`. For convenience we added two scripts for patch and minor version bumps (`bump:patch` and `bump:minor`).
Running one of those commands will increase the version and commit this change to master. It will also create a version tag (e. g. `v1.18.0`), which in turn triggers a build to run.
If this run is successful a Github release will be drafted. The name of the release will be the name of the tag.

You can now adjust the release notes on this tag. When you decide that you are ready for releasing, publish the release. This will trigger the release pipeline, that publishes all modules to npm.

### What to do when the build fails
You should only trigger a release, when the last master build succeeded. If the pipeline still fails for some reason, remove the tag locally and on Github, before fixing the issue.
Once the issue is fixed, you will have to create a tag manually. Creating the tag should trigger the process as described above.
