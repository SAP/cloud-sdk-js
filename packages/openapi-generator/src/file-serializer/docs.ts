import { EOL } from 'os';
import { documentationBlock } from '@sap-cloud-sdk/util';
import type {
  OpenApiApi,
  OpenApiNamedSchema,
  OpenApiObjectSchemaProperty,
  OpenApiOperation,
  OpenApiParameter,
  OpenApiRequestBody
} from '../openapi-types';

export function apiDocumentation(api: OpenApiApi, serviceName: string): string {
  return documentationBlock`
  Representation of the '${api.name}'.
  This API is part of the '${serviceName}' service.
  `;
}

export function operationDocumentation(
  operation: OpenApiOperation,
  operationResponseType: string
): string {
  const signature: string[] = [];
  if (operation.pathParameters.length) {
    signature.push(...getSignatureOfPathParameters(operation.pathParameters));
  }
  if (operation.requestBody) {
    signature.push(getSignatureOfBody(operation.requestBody));
  }
  if (operation.queryParameters.length) {
    signature.push(
      '@param queryParameters Object containing the query parameters.'
    );
  }
  signature.push(`@returns ${operationResponseType}`);

  return documentationBlock`
  ${getOperationDescriptionText(operation, operationResponseType)}
  
  ${signature.join(EOL)}
  `;
}

export function schemaDocumentation(schema: OpenApiNamedSchema): string {
  return documentationBlock`
  ${schema.description || `Representation of the '${schema.name}' schema`}
  `;
}

export function schemaPropertyDocumentation(
  schema: OpenApiObjectSchemaProperty
): string {
  return schema.description
    ? documentationBlock`${schema.description}` + EOL
    : '';
}

function getSignatureOfPathParameters(
  parameters: OpenApiParameter[]
): string[] {
  return parameters.map(
    (parameter, i) =>
      `@param ${parameter.name} ${
        parameter.description ||
        `Path parameter with the original name ${parameter.originalName}`
      }`
  );
}

function getSignatureOfBody(body: OpenApiRequestBody): string {
  return `@param body ${body.description || 'Request body'}`;
}

function getOperationDescriptionText(
  operation: OpenApiOperation,
  operationResponseType: string
): string {
  if (operation.description) {
    return operation.description;
  }

  return `Makes a ${operation.method} request to the '${operation.pathPattern}' endpoint and returns a '${operationResponseType}'`;
}
