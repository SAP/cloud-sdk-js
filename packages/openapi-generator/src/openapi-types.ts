import { OpenAPIV3 } from 'openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of a parsed OpenApi Service.
 */
export interface OpenApiDocument {
  apiName: string;
  operations: OpenApiOperation[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of an operation.
 */
export interface OpenApiOperation extends OpenAPIV3.OperationObject {
  operationName: string;
  method: string;
  pattern: string;
  requestBody?: OpenApiRequestBody;
  parameters: OpenApiParameter[];
}

const supportedMethods = {
  get: 'get',
  put: 'put',
  post: 'post',
  patch: 'patch',
  delete: 'delete',
  head: 'head',
  options: 'options'
} as const;

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Methods supported by OpenApi and SAP Cloud SDK.
 */
export type Method = typeof supportedMethods[keyof typeof supportedMethods];

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Get supported methods.
 * @returns Methods supported by OpenApi and SAP Cloud SDK.
 */
export function methods(): Method[] {
  return Object.values(supportedMethods);
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenatiation of a parameter for both queries and path parameters.
 */
export interface OpenApiParameter extends OpenAPIV3.ParameterObject {
  type: string;
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenation of the request body.
 */
export interface OpenApiRequestBody extends OpenAPIV3.RequestBodyObject {
  parameterName: string;
  parameterType: string;
}
