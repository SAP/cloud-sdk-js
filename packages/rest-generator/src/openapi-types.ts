import { $Refs } from '@apidevtools/swagger-parser';
import { OpenAPIV3 } from 'openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of a parsed OpenApi Service.
 */
export interface OpenApiDocument {
  apiName: string;
  serviceDirName: string;
  refs: $Refs;
  operations: OpenApiOperation[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of an operation.
 */
export interface OpenApiOperation extends OpenAPIV3.OperationObject {
  method: string;
  pattern: string;
  requestBody?: OpenApiRequestBody;
  parameters: OpenApiParameter[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Methods supported by OpenApi and SAP Cloud SDK.
 */
export enum Method {
  get = 'get',
  put = 'put',
  post = 'post',
  patch = 'patch',
  delete = 'delete',
  head = 'head',
  options = 'options'
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenatiation of a parameter for both queries and path parameters.
 */
export interface OpenApiParameter extends OpenAPIV3.ParameterObject {
  type: string;
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenation of the request body.
 */
export interface OpenApiRequestBody extends OpenAPIV3.RequestBodyObject {
  name: string;
  type: string;
}
