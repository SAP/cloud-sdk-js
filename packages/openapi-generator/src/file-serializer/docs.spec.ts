import { OpenApiOperation } from '../openapi-types';
import { apiDocumentation, operationDocumentation, schemaDocumentation } from './docs';

describe('docs', () => {
  function getOperation(): OpenApiOperation {
    return {
      response: { type: 'string' },
      method: 'GET',
      pathPattern: 'my/Api',
      pathParameters: [] as any,
      queryParameters: [] as any
    } as OpenApiOperation;
  }

  it('creates documentation for the api', () => {
    expect(
      apiDocumentation(
        { name: 'TestApi' } as any,
        'TestService'
      )
    ).toMatchSnapshot();
  });

  it('creates a description for operations if not present', () => {
    const operation = getOperation();
    expect(operationDocumentation(operation)).toMatchSnapshot();
  });

  it('uses the description for operations if present', () => {
    const operation = {
      ...getOperation(),
      description: 'This is my Operation.'
    };
    expect(operationDocumentation(operation)).toMatch(/This is my Operation/);
  });

  it('creates documentation with path parameters', () => {
    const operation = getOperation();
    operation.pathParameters = [
      { name: 'pathParameter1' },
      { name: 'pathParameter2' }
    ] as any;
    expect(operationDocumentation(operation)).toMatchSnapshot();
  });

  it('uses path parameter description if present', () => {
    const operation = getOperation();
    operation.pathParameters = [
      {
        name: 'pathParameter1',
        description: 'This is my parameter description'
      }
    ] as any;
    expect(operationDocumentation(operation)).toMatch(
      /This is my parameter description/
    );
  });

  it('creates documentation with query parameters object', () => {
    const operation = getOperation();
    operation.queryParameters = [
      { name: 'queryParameter1' },
      { name: 'queryParameter2' }
    ] as any;
    expect(operationDocumentation(operation)).toMatchSnapshot();
  });

  it('creates documentation with body parameter', () => {
    const operation = getOperation();
    operation.requestBody = { schema: { type: 'string' }, required: true };
    expect(operationDocumentation(operation)).toMatchSnapshot();
  });

  it('uses the body description if present', () => {
    const operation = getOperation();
    operation.requestBody = { schema: { type: 'string' },description:'My body description', required: true };
    expect(operationDocumentation(operation)).toMatch(/My body description/);
  });

  it('creates the signature in order path parameter, body, queryParameter and returns last', () => {
    const operation = getOperation();
    operation.pathParameters = [{ name: 'pathParameter1' }] as any;
    operation.requestBody = { schema: { type: 'string' }, required: false };
    operation.queryParameters = [
      { name: 'queryParameter1' },
      { name: 'queryParameter2' }
    ] as any;
    operation.requestBody = { schema: { type: 'string' }, required: true };
    expect(operationDocumentation(operation)).toMatch(
      /@param pathParameter1.*\s.*@param body.*\s.*@param queryParameters.*\s.*@returns/
    );
  });

  it('creates schema documentation',()=>{
    expect(schemaDocumentation({ name:'mySchema' } as any)).toMatchSnapshot();
  });

  it('uses the schema description documentation if present',()=>{
    expect(schemaDocumentation({ name:'mySchema',description:'My schmema description.' } as any)).toMatch(/My schmema description/);
  });
});
