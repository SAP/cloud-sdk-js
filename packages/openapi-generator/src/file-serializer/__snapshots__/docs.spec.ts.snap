// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docs creates a description for operations if not present 1`] = `
"/**
 * Create a request builder for execution of GET requests to the 'my/Api' endpoint.
 * @returns OpenApiRequestBuilder Use the execute() method to trigger the request.
 */"
`;

exports[`docs creates a schema property documentation 1`] = `
"/**
 * My propert
 */"
`;

exports[`docs creates documentation for the api 1`] = `
"/**
 * Representation of the 'TestApi'.
 * This API is part of the 'TestService' service.
 */"
`;

exports[`docs creates documentation with body parameter 1`] = `
"/**
 * Create a request builder for execution of GET requests to the 'my/Api' endpoint.
 * @param body Request body.
 * @returns OpenApiRequestBuilder Use the execute() method to trigger the request.
 */"
`;

exports[`docs creates documentation with path parameters 1`] = `
"/**
 * Create a request builder for execution of GET requests to the 'my/Api' endpoint.
 * @param pathParameter1 Path parameter.
 * @param path-parameter-2 Path parameter.
 * @returns OpenApiRequestBuilder Use the execute() method to trigger the request.
 */"
`;

exports[`docs creates documentation with query parameters object 1`] = `
"/**
 * Create a request builder for execution of GET requests to the 'my/Api' endpoint.
 * @param queryParameters Object containing the following keys: queryParameter1, queryParameter2.
 * @returns OpenApiRequestBuilder Use the execute() method to trigger the request.
 */"
`;

exports[`docs creates schema documentation 1`] = `
"/**
 * Representation of the 'mySchema' schema.
 */"
`;
